### PR TITLE
Shanghai EVM EIPs Tests: EIP-3651, EIP-3855, EIP-3860

### PR DESCRIPTION
**Pull request by marioevz**
## Tests Included
- EIP-3651
- EIP-3855
- EIP-3860

## `ethereum_test_tools` Changes
- Added `generators.py` under the `code` module, which shall include all code-generating classes and functions
- Relocated `CodeGasMeasure` to this new file
- Added `Initcode` class, which generates a legacy initcode for a given code to deploy
- Adds helper functions:
  - `ceiling_division`: Calculates the division of two integers and returns the ceiling, without using floating point.
  - `compute_create_address`, `compute_create2_address`: Computes the address of a created contract when using create or create2 opcodes.
  - `eip_2028_transaction_data_cost`: Calculates the gas cost of a transaction's data after EIP-2028.
  - `is_fork`: Checks whether a fork is equal or greater than the specified fork.
- Adds `Shanghai` fork
- Print extra debugging information for better clarity when a test case fails

### ToDos
Rebase withdrawals on top of this branch for release.

### CODE REVIEW

File: fillers/eips/eip3860.py [hunk 1]
```diff
@@ -0,0 +1,704 @@
...
+    return cost
+
+
+def calculate_create_tx_execution_cost(
+    initcode: Initcode,
+    eip_3860_active: bool,
+) -> int:
+    """
+    Calculates the minimum total execution gas cost of a transaction that
+    contains initcode and creates a contract for the contract to be
```
Comment (lightclient) [inline fillers/eips/eip3860.py:90]:
```suggestion
    contains initcode and creates a contract
```

File: fillers/eips/eip3860.py [hunk 2]
```diff
@@ -0,0 +1,704 @@
...
+    successfully deployed
```
Comment (lightclient) [inline fillers/eips/eip3860.py:91]:
```suggestion
```

**Review by lightclient — COMMENTED** [review]
Mainly a few nits; overall looks good though. Nice work.

File: fillers/eips/eip3855.py [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+    yield StateTest(
+        env=env, pre=pre, post=post, txs=[tx], name="push0_key_sstore"
+    )
+
+    """
+    Test case 2: Fill stack with PUSH0, then OR all values and save using
+    SSTORE
+    """
+    pre[to_address(0x100)] = Account(
+        code=bytes(
```
[the remark this answers is not in the document]
Comment (marioevz) [reply fillers/eips/eip3855.py]: I changed the format a bit, I think it looks nicer while preserving the format overall.

File: fillers/eips/eip3651.py [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+        number=1,
+        timestamp=1000,
+    )
+
+    # List of opcodes that are affected by
+    gas_measured_opcodes: Dict[str, CodeGasMeasure] = {
+        "EXTCODESIZE": CodeGasMeasure(
+            code=bytes(
+                [
+                    0x41,  # addr: COINBASE
```
[the remark this answers is not in the document]
Comment (marioevz) [reply fillers/eips/eip3651.py:175]: Definitely, I will look into this for a follow up PR, thanks!

**Review by lightclient — APPROVED** [review]

**Pull request merged**